[swift-inspect] implement basic GNU/Linux support

andrurogerz opened the pull request:
## Purpose
Implement support for Linux in the `swift-inspect` tool. Omits support for heap iteration, for which there is no standard GNU/Linux mechanism (see #63576). Currently supports only 64-bit processes, though minimal changes are required to support 32-bit (elf32 parsing is included).

## Overview
* Extends the existing swift-inspect implementation to support Linux along side the existing Darwin and Windows implementations
* Excludes `DumpArray` functionality from Linux builds (does not work without heap iteration)
* Adds a new `SwiftInspectLinux` build target with the Linux-specific implementation:
  * Remote process reading using `process_vm_readv`
  * Memory map parsing of `/proc/<pid>/maps`
  * Aux vector parsing of `/proc/<pid>/auxv` used to locate debugger link map
  * Remote process debugger link map parsing to generate the list of modules loaded within the target process
  * Minimal elf32 and elf64 parsing required to read string and symbol tables from executable files
* Adds a new `SwiftInspectLinux/SystemHeaders` target to pull-in the required Linux C headers
* Updates `README.md` to reflect Linux support

## Background
I have a locally working implementation of `swift-inspect` for Android, and I realized that, other than the heap iteration implementation, it is not Android-specific. This PR extracts the GNU/Linux functionality into stand-alone Linux support.

Linux support has one glaring omission: heap iteration is not supported. There is no standard GNU/Linux heap iteration interface. Further, Linux programs often use alternative heap implementations (scudo, jemalloc, mimalloc, etc) so parsing raw heap structures gets complex (see see #63576). Fortunately, only the `dump-arrays` and `dump-concurrency` commands rely on heap iteration-- the other commands work fine without it.

## Validation
* Built with `swift build` on Fedora Linux (6.11.8-200.fc40.x86_64)
* Manually tested all commands against a trivial command-line Swift app run with `SWIFT_DEBUG_ENABLE_METADATA_ALLOCATION_ITERATION=1` and `SWIFT_DEBUG_ENABLE_METADATA_BACKTRACE_LOGGING=1`
* Confirmed backtraces are printed with the `--backtrace` and `--backtrace-long` arguments.

## Note to reviewers
This is my first non-trivial Swift project, so feedback on Swift style/patterns/etc is appreciated.

andrurogerz committed:
[linux] swift-inspect support for Linux

andrurogerz committed:
[linux] add Linux build instructions to README.txt

andrurogerz committed:
[linux] exclude dump-array support

compnerd commented on tools/swift-inspect/README.md:
```diff
@@ -16,6 +16,14 @@ In order to build on Windows, some additional parameters must be passed to the b
+In order to build on Linux, some additional parameters must be passed to the build tool to loace the necessary includes and libraries
+
+~~~
+swift build -Xswiftc -I$(git rev-parse --show-toplevel)/include/swift/SwiftRemoteMirror -Xswiftc -lswiftRemoteMirror
```
Can we use `systemLibrary` to avoid the `-Xswiftc -lswiftRemoteMirror`? I think that we should also avoid the `-Xswiftc -lswiftRemoteMirror` at least in favour of `-Xlinker -lswiftRemoteMirror` or `-Xclang-linker -lswiftRemoteMirror`.

andrurogerz replied:
This should definitely be `-Xlinker -lswiftRemoteMirror` instead of using `-Xswiftc`.

I don't see how I can add a dependency on `SwiftRemoteMirror` as a `systemLibrary` since it is in a different location of the source tree.

compnerd commented on tools/swift-inspect/Sources/SwiftInspectLinux/AuxVec.swift:
```diff
@@ -0,0 +1,73 @@
+import LinuxSystemHeaders
+
+internal class AuxVec {
+  enum Error: Swift.Error { case FileReadFailure(_ filePath: String) }
```
Why is this part of `AuxVec`?

andrurogerz replied:
eh, I have it throw that error when `load()` fails but I can just as easily replace `throws` with an optional result

compnerd commented on tools/swift-inspect/Sources/SwiftInspectLinux/AuxVec.swift:
```diff
@@ -0,0 +1,73 @@
+  enum Error: Swift.Error { case FileReadFailure(_ filePath: String) }
+
+  // enum values must match the constants defined in usr/include/linux/auxv.h
+  enum Tag: UInt64 {
```
Can we use a module map to import `linux/auxv.h` instead? This seems unnecessarily fragile.

andrurogerz replied:
Sure, I'll drop the `enum`.

compnerd commented on tools/swift-inspect/Sources/SwiftInspectLinux/AuxVec.swift:
```diff
@@ -0,0 +1,73 @@
+    let fileHandle = try FileHandle(forReadingFrom: URL(fileURLWithPath: filePath))
+    defer { fileHandle.closeFile() }
+
+    guard let data = try fileHandle.readToEnd() else { throw Error.FileReadFailure(filePath) }
```
Can you change this to use [`Data(contentsOf:options:)`](https://developer.apple.com/documentation/foundation/nsdata/1407864-init)?

andrurogerz replied:
I wrote it that way originally, but it always returns empty `Data` object. I also tried `String(contentsOfFile:` and that gave back an empty string. I didn't look at the implementations, but I suspect the issue is because those methods `stat` the file to get the size whereas `FileHandle.readToEnd` just calls `read` until it gets eof.

andrurogerz replied:
And to be totally clear: `stat(2)` on the file `/proc/<pid>/auxvec` reports a size of 0 bytes. You can confirm this with `ls -l /proc/self/` and observe most of the procfs files are 0 bytes.

compnerd commented on tools/swift-inspect/Sources/SwiftInspectLinux/AuxVec.swift:
```diff
@@ -0,0 +1,73 @@
+    guard let data = try fileHandle.readToEnd() else { throw Error.FileReadFailure(filePath) }
+
+    // aux vector is an array of 8-byte pairs in a 64-bit process
+    assert(process.elfFile.isElf64, "only 64-bit processes are supported")
```
Hmm, this can potentially be compiled out, can we change this to `precondition`?

andrurogerz replied:
Good catch; I meant to replace the assert.

compnerd commented on tools/swift-inspect/Sources/SwiftInspectLinux/AuxVec.swift:
```diff
@@ -0,0 +1,73 @@
+    for (rawTag, value) in auxVec {
+      guard let tag = Tag(rawValue: rawTag) else { continue }
+      entries[tag] = value
+    }
```
I think that it would be better to sink this into the `withUnsafeBytes` view and return the dictionary directly.

compnerd commented on tools/swift-inspect/Sources/SwiftInspectLinux/ElfDyn.swift:
```diff
@@ -0,0 +1,32 @@
+  static var symbolSize: Int { MemoryLayout<Elf32_Dyn>.size }
+  var tag: Int64 { Int64(self.d_tag) }
+  var val: UInt64 { UInt64(self.d_un.d_val) }
+}
```
Please define these types in a C header as we cannot guarantee the type layout in Swift.

compnerd commented on tools/swift-inspect/Sources/SwiftInspectLinux/ElfEhdr.swift:
```diff
@@ -0,0 +1,68 @@
+  var shentsize: UInt16 { self.e_shentsize }
+  var shnum: UInt16 { self.e_shnum }
+  var shstrndx: UInt16 { self.e_shstrndx }
+}
```
Likewise

compnerd commented on tools/swift-inspect/Sources/SwiftInspectLinux/ElfFile.swift:
```diff
@@ -0,0 +1,184 @@
+    case FileReadFailure(_ filePath: String, offset: UInt64, size: UInt64)
+    case FileNotElfFormat(_ filePath: String)
+    case MalformedElfFile(_ filePath: String, description: String = "")
+  }
```
Please use `ELF` instead of `Elf`. However, some of these feel like they can be replaced with Foundation error values?

andrurogerz replied:
I will change `Elf` -> `ELF` in the errors. I plan to keep the `Elf` naming convention for data structures to match definitions in `elf.h`. Let me know if you disagree with that.

More generally, I made my own error types because I didn't see general purpose ones. Digging a bit further now, I still only see `NSError`, `CFError`, and `CocoaError`, none of which seems fit for purpose here. I could probably use `POSIXError` in a few places. Am I overlooking something obvious?

compnerd commented on tools/swift-inspect/Sources/SwiftInspectLinux/ElfFile.swift:
```diff
@@ -0,0 +1,184 @@
+
+    guard let file = try? FileHandle(forReadingFrom: URL(fileURLWithPath: filePath)) else {
+      throw Error.FileOpenFailure(filePath)
+    }
```
Does it make sense to use `Data(contentsOf:options:)` to use a mmap'ed view?

andrurogerz replied:
I can if you want that, but I prefer explicit file i/o over `mmap` because it allows explicit error handling. They both share pages in the page cache so isn't going to perform significantly better than the other.

compnerd commented on tools/swift-inspect/Sources/SwiftInspectLinux/ElfFile.swift:
```diff
@@ -0,0 +1,184 @@
+
+        // sym.name is a byte offset into the string table
+        guard let strStart = Int(exactly: sym.name), strStart < strTable.count else {
+          throw Error.MalformedElfFile(
```
Is this safe? There is no guarantee that there is a string table in the binary.

andrurogerz replied:
This bit is safe because it won't even get here if there's no string table to load, but its a fair point that an Elf file may not have symbol or string table sections.

andrurogerz replied:
fwiw, I implemented the `loadSymbolTable` function to return an empty symbol table if it doesn't find a symtab or dynsym section to load symbols from

compnerd commented on tools/swift-inspect/Sources/SwiftInspectLinux/ElfPhdr.swift:
```diff
@@ -0,0 +1,50 @@
+  var memsz: UInt64 { UInt64(self.p_memsz) }
+  var flags: UInt32 { self.p_flags }
+  var align: UInt64 { UInt64(self.p_align) }
+}
```
Please define these in a C header

andrurogerz replied:
Can I move this to a C header? It is an extension to the existing `Elf32_Phdr` C struct which is imported via `elf.h`. My goal with this file (and the other `Elf*.swift` files) was to create a generic way of accessing both elf32 and elf64 structs, and this protocol + extending of the C structs pattern is what I came up with. Is there a better way to achieve the same?

mikeash replied:
This looks good, I think @compnerd misread this as defining the type rather than just extending it.

compnerd commented on tools/swift-inspect/Sources/SwiftInspectLinux/ElfShdr.swift:
```diff
@@ -0,0 +1,56 @@
+  var info: UInt32 { self.sh_info }
+  var addralign: UInt64 { self.sh_addralign }
+  var entsize: UInt64 { self.sh_entsize }
+}
```
Likewise

compnerd commented on tools/swift-inspect/Sources/SwiftInspectLinux/ElfSym.swift:
```diff
@@ -0,0 +1,44 @@
+  var shndx: UInt16 { self.st_shndx }
+  var value: UInt64 { self.st_value }
+  var size: UInt64 { self.st_size }
+}
```
Likewise

compnerd commented on tools/swift-inspect/Sources/swift-inspect/Operations/DumpArray.swift:
```diff
@@ -10,6 +10,8 @@
 //
 //===----------------------------------------------------------------------===//
 
+#if os(iOS) || os(macOS) || os(tvOS) || os(watchOS) || os(Windows)
```
Maybe compress this to `!os(Linux)`?

andrurogerz committed:
[linux] define _GNU_SOURCE for process_vm_readv

andrurogerz committed:
address some code review comments

andrurogerz committed:
specify -Xlinker instead of -Xswiftc in README.md

al45tair commented:
The ELF code should probably not be here; we should share (and if need be extend) the existing ELF reading code in `stdlib/public/Backtracing` (which is going to move to `stdlib/public/RuntimeModule` when I get enough time to finish working on that).

mikeash commented on tools/swift-inspect/README.md:
```diff
@@ -16,6 +16,14 @@ In order to build on Windows, some additional parameters must be passed to the b
 
+#### Linux
+
+In order to build on Linux, some additional parameters must be passed to the build tool to loace the necessary includes and libraries
```
Typo, `loace` -> `locate`.

mikeash commented on tools/swift-inspect/Sources/SwiftInspectLinux/AuxVec.swift:
```diff
@@ -0,0 +1,49 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
```
Can bump the end year to 2024 here.

mikeash commented on tools/swift-inspect/Sources/SwiftInspectLinux/AuxVec.swift:
```diff
@@ -0,0 +1,49 @@
+  // loads the auxiliary vector for a process
+  public static func load(for process: Process) -> [Int32 : UInt64]? {
+    let filePath = "/proc/\(process.pid)/auxv"
+    guard let fileHandle = FileHandle(forReadingAtPath: filePath) else { return nil }
```
Might be good to have a brief comment here about how `Data(contentsOf:)` doesn't work due to the `stat` issue. I can imagine someone trying to "improve" this and being sad.

mikeash commented on tools/swift-inspect/Sources/SwiftInspectLinux/AuxVec.swift:
```diff
@@ -0,0 +1,49 @@
+    guard let fileHandle = FileHandle(forReadingAtPath: filePath) else { return nil }
+    defer { fileHandle.closeFile() }
+
+    guard let data = try? fileHandle.readToEnd(), data.count > 0 else { return nil }
```
Should a count of 0 produce `nil` versus an empty dictionary? Not sure if this aux vector thing is ever supposed to be empty.

andrurogerz replied:
auxvec can never be empty, so I think in this case failing on an empty file is reasonable. Either way works, though.

mikeash replied:
Sounds good, if it should never be empty then `nil` makes sense to me.

mikeash commented on tools/swift-inspect/Sources/SwiftInspectLinux/MemoryMap.swift:
```diff
@@ -0,0 +1,80 @@
+      } else if entry.endAddr <= addr {
+        lowerBound = currentIndex + 1
+      } else {
+        assert(addr >= entry.startAddr)
```
I'd suggest `precondition` instead of `assert` as long as performance is acceptable in release builds (which I expect should be the case here).

mikeash commented on tools/swift-inspect/Sources/SwiftInspectLinux/Process.swift:
```diff
@@ -0,0 +1,79 @@
+  public func readArray<T>(address: UInt64, upToCount: UInt) throws -> [T] {
+    let maxSize = upToCount * UInt(MemoryLayout<T>.stride)
+    let array: [T] = Array(unsafeUninitializedCapacity: Int(upToCount)) { buffer, initCount in
+      var local = iovec(iov_base: buffer.baseAddress!, iov_len: Int(maxSize))
```
I think `baseAddress` is allowed (but not required) to be `nil` if `upToCount` is `0`. Might be good to have an early check for `upToCount == 0` and return an empty array, or an error if you think `0` is never going to be a valid request here.

andrurogerz replied:
Great catch. I will add that check.

mikeash commented on tools/swift-inspect/Sources/SwiftInspectLinux/SymbolCache.swift:
```diff
@@ -0,0 +1,74 @@
+    self.symbolLookup = symbolLookup
+  }
+
+  // find the address range for the requetsed symbol
```
Typo, `requetsed` -> `requested`

mikeash commented on tools/swift-inspect/Sources/swift-inspect/LinuxRemoteProcess.swift:
```diff
@@ -0,0 +1,155 @@
+
+        // copy the string from the remote proces to get its length
+        guard let string = try? process.process.readString(address: address),
+          let len = UInt64(exactly: string.count)
```
The caller will expect the count in bytes, so this will be short for non-ASCII UTF-8 data, which I think is a realistic possibility if people use non-ASCII in type names or something. `string.utf8.count` should be correct, although I'm mildly worried about normalization coming along and ruining our day. I don't _think_ `String` eagerly normalizes its data though.

andrurogerz replied:
Ah, ok. Yeah it makes sense the caller wants byte count rather than character count. I will look into it a bit more and use `String.utf8.count` if that looks safe. I could also add a method to just read null-terminated raw bytes.

mikeash replied:
Raw bytes might make more sense, given that this is a C interface with no defined encoding. A `Array<UInt8>` that ends in a `0` would be a decent representation of that.

mikeash commented on tools/swift-inspect/Sources/swift-inspect/LinuxRemoteProcess.swift:
```diff
@@ -0,0 +1,155 @@
+      self.processIdentifier = processId
+      do {
+        let path = "/proc/\(processId)/cmdline"
+        self.processName = try String(contentsOfFile: path, encoding: .ascii)
```
Should this be `.utf8`? I don't think any encoding beyond bag-of-bytes is entirely correct, but UTF-8 would stand a chance of doing the right thing for a non-ASCII filename.

mikeash commented on tools/swift-inspect/Sources/SwiftInspectLinux/ElfFile.swift:
```diff
@@ -0,0 +1,184 @@
+import Foundation
+import LinuxSystemHeaders
+
+class ElfFile {
```
You might consider making this generic on something like `<T: ElfEhdr>` and having a small wrapper function to decide which one to instantiate. That would let you remove the `isElf64 ? ... : ...` checks and the use of existentials for things like `ElfSym`, which might be marginally more efficient and nicer to read. Just an idea, definitely not mandatory here.

mikeash reviewed (commented):
I left a few comments, didn't look too deeply into the Linux-specific parts as I don't know that stuff, but it looks great overall.

andrurogerz committed:
PR feedback: fix typo in README.md

andrurogerz committed:
PR feedback: update copyright to 2024

andrurogerz committed:
consistent use of /proc

andrurogerz committed:
PR feedback: use precondition in place of assert

andrurogerz committed:
PR feedback: short-circuit return on readArray of 0 items

andrurogerz committed:
PR feedback: add Process.readRawString to get byte count in GetStringLengthFunction

andrurogerz committed:
minimize ELF parsing support to ELF64 only

andrurogerz committed:
PR feedback: rework error definitions

andrurogerz committed:
PR feedback: fix typo

andrurogerz commented:
Thank you @compnerd, @mikeash, and @al45tair for the feedback. I believe I have addressed the bulk of your comments.

I have revised the ELF parsing code to be as small as possible: it contains only what is required for loading symbol tables from ELF64 files. Dropping ELF32 (which wasn't supported yet for other reasons) simplified it quite a bit. Since there is no clear way to share the ELF parsing code with the `swift/stdlib/public/Backtracing` library right now, I have added a TODO to the new `ElfFile.swift` file indicating it should be replaced in the future (specifically if/when the internal ELF code is refactored to a public location.

mikeash reviewed (approved):
One more tiny suggestion, looking good!

andrurogerz committed:
properly override the Free property in LinuxRemoteProcess

andrurogerz committed:
PR Feedback: make ProcFS an enum instead of a class

andrurogerz committed:
formatting fixes and missed copyright date updates

andrurogerz commented:
@al45tair do you have any additional thoughts on this PR? It seems worth the minimal code duplication of ELF parsing (I've reduced it to a single 160 line source file) to enable swift-inspect support for Linux and Android (coming next). It can eventually be updated to share the ELF code with the backtrace library if that gets extracted to its own foundation library.

al45tair commented:
@andrurogerz I'm OK with it if Mike is; the duplication is something we should aim to remove in the medium to long term, to avoid having to maintain two sets of ELF code, but I have no objection to this going in.

al45tair commented:
@swift-ci Please smoke test

DougGregor commented:
@swift-ci Please smoke test

andrurogerz commented:
@swift-ci please test

compnerd commented on tools/swift-inspect/Sources/SwiftInspectLinux/SymbolCache.swift:
```diff
@@ -0,0 +1,74 @@
+    self.linkMap = try LinkMap(for: process)
+    var symbolLookup: SymbolLookup = [:]
+    for linkMapEntry in linkMap.entries {
+      guard FileManager.default.fileExists(atPath: linkMapEntry.moduleName) else { continue }
```
This feels like a TOCTOU issue.

andrurogerz replied:
I suppose it is. I added this check so we don't try to open the "[vdso]" entry in the linkmap, but I could just as easily detect and skip that OR catch a not found exception from `try ElfFile(...)`.

compnerd commented on tools/swift-inspect/Sources/SwiftInspectLinux/ProcFS.swift:
```diff
@@ -0,0 +1,36 @@
+    // handles these cases.
+    guard let fileHandle = FileHandle(forReadingAtPath: filePath) else { return nil }
+    defer { fileHandle.closeFile() }
+    guard let data = try? fileHandle.readToEnd() else { return nil }
```
Why not use the `Data(contentsOf:options:)` method instead of the raw filehandle?

andrurogerz replied:
See the comment above. It doesn't work on many files under `/proc`.

compnerd commented on tools/swift-inspect/Sources/SwiftInspectLinux/ElfFile.swift:
```diff
@@ -0,0 +1,162 @@
+  public init(filePath: String) throws {
+    self.filePath = filePath
+
+    let file = try FileHandle(forReadingFrom: URL(fileURLWithPath: filePath))
```
Why not use `Data(contentsOf:options:)` and mmap the file and read the first n bytes?

andrurogerz replied:
I typically prefer explicit file I/O so you can handle failures, but I am happy to rewrite this class using `mmap` file if you prefer that.

andrurogerz replied:
Reworked the `ElfFile` implementation to use `Data(options: .alwaysMapped` instead of file seek/read.

andrurogerz committed:
PR feedback

andrurogerz committed:
use memory-mapped IO instead of file reads

compnerd commented on tools/swift-inspect/Sources/SwiftInspectLinux/ElfFile.swift:
```diff
@@ -0,0 +1,144 @@
+  public init(filePath: String) throws {
+    self.filePath = filePath
+
+    let fileData = try Data(contentsOf: URL(fileURLWithPath: filePath), options: .alwaysMapped)
+    self.fileData = fileData
```
```suggestion
    self.fileData = try Data(contentsOf: URL(fileURLWithPath: filePath), options: .alwaysMapped)
```

compnerd commented on tools/swift-inspect/Sources/SwiftInspectLinux/ElfFile.swift:
```diff
@@ -0,0 +1,144 @@
+    let fileData = try Data(contentsOf: URL(fileURLWithPath: filePath), options: .alwaysMapped)
+    self.fileData = fileData
+
+    let identLen = Int(EI_NIDENT)
+    let identMagic = String(bytes: fileData[0..<identLen].prefix(Int(SELFMAG)), encoding: .utf8)
+    guard identMagic == ELFMAG else {
+      throw ELFError.notELF64(filePath)
+    }
```
```suggestion
    let ident = fileData.prefix(upTo: Int(EI_NIDENT))

    guard String(bytes: ident.prefix(Int(SELFMAG)), encoding: .utf8) == ELFMAG else {
      throw ELFError.notELF64(filePath)
    }
```

compnerd commented on tools/swift-inspect/Sources/SwiftInspectLinux/ElfFile.swift:
```diff
@@ -0,0 +1,144 @@
+      throw ELFError.notELF64(filePath)
+    }
+
+    let identClass = fileData[Int(EI_CLASS)]
```
Should this not operate on the identity buffer rather than indexing into the file data? It _just_ happens that the base indicies are equal and this works.

compnerd commented:
@swift-ci please smoke test

andrurogerz committed:
PR feedback in ElfFile.init

compnerd commented:
@swift-ci please smoke test

andrurogerz commented:
Hm, the Windows test failure doesn't look like it could have been caused by this PR

DougGregor commented:
@swift-ci please smoke test Windows
